torch/_dynamo/_trace_wrapped_higher_order_op.py

```diff
@@ -0,0 +1,111 @@
+import torch.utils._pytree as pytree
+from torch._C import DispatchKey
+from torch._higher_order_ops.utils import autograd_not_implemented
+
+from torch._ops import HigherOrderOperator
+from torch._subclasses import FakeTensorMode
+
+from torch.fx.experimental.proxy_tensor import ProxyTorchDispatchMode, track_tensor_tree
+from torch.utils._python_dispatch import _get_current_dispatch_mode
+
+
+__all__ = ["trace_wrapped"]
+
+
+# Trace wrapped is a higher order op meant for both invoking a bound function,
+# and for registering it as a call_function.
+# This allows us to re-enter dynamo during compiled autograd to trace (or graph break)
+# the functions as needed. While there is nothing backward specific about this op, the way it is written means
+# we can support functions in backward with complex python. It can be thought of as an allow_in_graph
+# for our aten graph. If we were to not do this, the functions would get inlined into their composing aten ops,
+# and we would lose the python state mutation.
```

Comment:
Here is a proposed rewrite of the top level comment:

`trace_wrapped(*args, fn)` is equivalent to `fn(*args)`, but with a twist: if you `make_fx` trace through this call, we will not actually trace into fn; instead, we will directly insert it as a `call_function` to `fn` in the graph. (Unlike `make_fx`, Dynamo WILL inline into fn.) You can think of this as a one off `allow_in_graph` equivalent for proxy tensor tracing.

Because proxy tensor tracing does not actually run the function, there are requirements on the behavior of fn. We are still figuring it out, but here is the current state:

* fn can only take a single argument
* fn must return a new tensor with the same metadata as the original tensor (e.g., `empty_like(input)` is a permissible implementation of fn). This is verified via an extra assert that is inserted into the traced graph.
* fn MAY have side effects, but it MAY NOT perform metadata mutation on other tensors participating in proxy tensor tracing (it MAY mutate other tensors, it MAY mutate Python state)

These requirements stem from the requirement that we need to continue performing proxy tensor tracing, which assumes accurate fake tensor metadata, without actually running fn. In the future, we may allow for a "meta" function associated with fn to allow for more interesting input-output patterns.

Note that tensors / Python state are allowed to be mutated. This is relaxed constraint is not always sound, but it is sound for backward tracing with fake tensors as it takes place in AOTAutograd, as the backward pass is guaranteed not to depend on concrete tensor values (via fake tensor) or Python state (because the autograd engine doesn't depend on Python).

The intended use case for this function is to allow AOTAutograd to defer complex backward hooks to compiled autograd. AOTAutograd performs a make_fx trace which preserves the function call as is in the graph, and only when we Dynamo through the backward graph in compiled autograd do we inline into the function.

Reply:
Sure, this is better. Thanks for rewriting it. eg: zeros_like(input) I suppose

Reply:
if you are using may and must as https://www.rfc-editor.org/rfc/rfc2119 - let's use SHOULD and MUST ;) 

Thank you again.